### PR TITLE
Changes to ensure forward progress in fullnode sync

### PR DESCRIPTION
**Pull request by mystenmark**
Stacked on #4154 - top 6 commits are new.

- Store votes in DB instead of memory to further ensure forward progress0
- Early exit instead of waiting forever when a tx is not yet final.
- Actively download parent certs when necessary, to ensure forward progress.
- Time out sync tasks so they can't live forever
- Refactor how we store certs/effects in the NodeSyncStore. The goal of this refactoring was:
   - make sure we never download things multiple times - this is important to allowing us to make progress in light of the timeouts.
   - make sure we never store an un-verified effects to the db. Only effects that we know to be true are written.

### CODE REVIEW

File: crates/sui-storage/src/node_sync_store.rs [hunk 1]
```diff
@@ -15,25 +18,55 @@ use typed_store_macros::DBMapUtils;
...
+
 /// NodeSyncStore store is used by nodes to store downloaded objects (certs, etc) that have
 /// not yet been applied to the node's SuiDataStore.
 #[derive(DBMapUtils)]
 pub struct NodeSyncStore {
-    /// Certificates/Effects that have been fetched from remote validators, but not sequenced.
     certs_and_fx: DBMap<TransactionDigest, (CertifiedTransaction, SignedTransactionEffects)>,
 
+    /// Certificates that have been fetched from remote validators, but not sequenced.
+    certs: DBMap<TransactionDigest, CertifiedTransaction>,
```
Comment (lxfind) [inline crates/sui-storage/src/node_sync_store.rs:31]: Why do we need both certs and certs_and_fx? Also, is there a reason we cannot just use the certs table in authority store?

Comment (mystenmark) [reply crates/sui-storage/src/node_sync_store.rs:31]: Sorry, we don't - meant to delete the previous table.

We could use the certs table in the authority store for this - I was a little worried about the possibility of confusion when there is some other random process mutating the authority db, but it wouldn't be a correctness issue.

Comment (mystenmark) [reply crates/sui-storage/src/node_sync_store.rs]: Cleaned all of this up.

File: crates/sui-core/src/node_sync/node_state.rs [hunk 2]
```diff
@@ -371,69 +343,113 @@ where
...
-        digest: &TransactionDigest,
-    ) -> SuiResult {
-        trace!(?digest, "validator pending execution requested");
+        cert: &CertifiedTransaction,
+    ) -> SuiResult<SignedTransactionEffects> {
+        let digest = cert.digest();
+        match self.node_sync_store.get_effects(digest)? {
+            Some(effects) => Ok(effects),
+            None => {
+                let effects = self.aggregator.execute_cert_to_true_effects(cert).await?;
```
Comment (lxfind) [inline crates/sui-core/src/node_sync/node_state.rs:354]: The `execute_cert_to_true_effects` API in AuthorityAggregator is a bit strange to me. It returns the last signed effect when validity is reached. The signature is probably useless. Raw effects should be sufficient.
Similarly we should just store raw effects in the effects table.

**Review by lxfind — APPROVED** [review]

File: crates/sui-core/src/node_sync/node_state.rs (hunk 2 above)
Comment (mystenmark) [reply crates/sui-core/src/node_sync/node_state.rs:354]: Agreed, will fix. I just did it this way because it was the most convenient thing.

**Pull request merged**